src/ATokenCheckpointed.sol

```diff
@@ -0,0 +1,307 @@
+  function proposalDeadline(uint256 proposalId) external view returns (uint256);
+  function castVoteWithReasonAndParams(
+    uint256 proposalId,
+    uint8 support,
+    string calldata reason,
+    bytes memory params
+  ) external returns (uint256);
+}
+
+interface IVotingToken {
+  function transfer(address to, uint256 amount) external returns (bool);
+  function transferFrom(address from, address to, uint256 amount) external returns (bool);
+  function delegate(address delegatee) external;
+  function getPastVotes(address account, uint256 blockNumber) external view returns (uint256);
+}
+
+contract ATokenCheckpointed is AToken {
+  using WadRayMath for uint256;
+  using SafeCast for uint256;
+  using GPv2SafeERC20 for IERC20;
+  using Checkpoints for Checkpoints.History;
+
+  /// @notice The voting options corresponding to those used in the Governor.
+  enum VoteType {
+    Against,
+    For,
+    Abstain
+  }
+
+  /// @notice Data structure to store vote preferences expressed by depositors.
+  struct ProposalVote {
+    uint128 againstVotes;
+    uint128 forVotes;
+    uint128 abstainVotes;
+  }
+
+  /// @notice The number of blocks prior to the proposal deadline within which
+  /// `castVote` may be called. Prior to this window, `castVote` will revert so
+  /// as to give users time to call `expressVote` before votes are sent to the
+  /// governor contract.
+  uint32 public immutable CAST_VOTE_WINDOW;
+
+  /// @notice Map proposalId to an address to whether they have voted on this proposal.
+  mapping(uint256 => mapping(address => bool)) private proposalVotersHasVoted;
+
+  /// @notice Map proposalId to whether or not this contract has cast votes on it.
+  mapping(uint256 => bool) public hasCastVotesOnProposal;
+
+  /// @notice Map proposalId to vote totals expressed on this proposal.
+  mapping(uint256 => ProposalVote) public proposalVotes;
+
+  /// @notice The governor contract associated with this governance token. It
+  /// must be one that supports fractional voting, e.g. GovernorCountingFractional.
+  IFractionalGovernor public immutable GOVERNOR;
+
+  /// @notice Mapping from address to stored (not rebased) balance checkpoint history.
+  mapping(address => Checkpoints.History) private balanceCheckpoints;
+
+  /// @notice History of total underlying asset balance.
+  Checkpoints.History private totalDepositCheckpoints;
+
+  /// @dev Constructor.
+  /// @param _pool The address of the Pool contract
+  /// @param _governor The address of the flex-voting-compatible governance contract.
+  /// @param _castVoteWindow The number of blocks that users have to express
+  /// their votes on a proposal before votes can be cast.
+  constructor(IPool _pool, address _governor, uint32 _castVoteWindow) AToken(_pool) {
+    GOVERNOR = IFractionalGovernor(_governor);
+    CAST_VOTE_WINDOW = _castVoteWindow;
+  }
+
+  // TODO Is there a better way to do this? It cannot be done in the constructor
```

Comment:
nit: it might be more idiomatic to put this in an initializer

Reply:
Good point!

The trouble is that most aTokens will (presumably) already have been initialized if Aave ends up adopting this. So the thinking had been that we'd probably need to do this during an upgrade call. We were tracking that work here: https://github.com/ScopeLift/flexible-voting/issues/14

But you're making me realize that I should update this comment to clarify

Reply:
Actually, I think you're right, we should add this to the initializer so that new aTokens self-delegate. Existing aTokens will still need to be upgraded, though.

Great catch!

Reply:
Dang, `initialize` [isn't virtual](https://github.com/aave/aave-v3-core/blob/f3e037b3638e3b7c98f0c09c56c5efde54f7c5d2/contracts/protocol/tokenization/AToken.sol#L53-L62). This is another upstream change we'll need to make to aave.

For now, I'll add this to our existing ticket for upstream aave changes (https://github.com/ScopeLift/flexible-voting/issues/16) and update the comment accordingly.

Reply:
Comment added in https://github.com/ScopeLift/flexible-voting/commit/9d8171bb0a91c502d19cd1fb1180a7440f578c7f

Reply:
We have to make `aToken.initialize` both public and virtual. I did that locally, and then updated the initializer here: d7a6bf5

This will fail in CI (because it doesn't have my local Aave changes) but should pass if/when Aave merges the change

Reply:
Aave merged our PR earlier today and I've bumped the dependency in [e3ab13b](https://github.com/ScopeLift/flexible-voting/pull/21/commits/e3ab13b2366ec9238d037d49d2e5016502db31f2). So this should be taken care of now